In [1]:
import pandas as pd
import numpy as np
import json


In [2]:
df_archive = pd.read_csv('twitter-archive-enhanced.csv')
df_archive.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

In [3]:
df_class = pd.read_csv('twitter_image_class.csv')
df_class.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

In [4]:
df_status = pd.read_csv('twitter_status_dogs.csv')
df_status.head(3)

tweet_id                                         text_check  \
0  892420643555336193  This is Phineas. He's a mystical boy. Only eve...   
1  892177421306343426  This is Tilly. She's just checking pup on you....   
2  891815181378084864  This is Archie. He is a rare Norwegian Pouncin...   

                                  dog_picture_url  \
0  http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   
1  http://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2  http://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   

                                       tweet_web_url  favourite_count  \
0  https://twitter.com/dog_rates/status/892420643...            37434   
1  https://twitter.com/dog_rates/status/892177421...            32164   
2  https://twitter.com/dog_rates/status/891815181...            24251   

   retweet_count retweet_status_id_api  
0           8138                  None  
1           6029                  None  
2           3991                  None

In [5]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [6]:
df_class.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [7]:
df_status.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2335 entries, 0 to 2334
Data columns (total 7 columns):
tweet_id                 2335 non-null int64
text_check               2335 non-null object
dog_picture_url          2335 non-null object
tweet_web_url            2335 non-null object
favourite_count          2335 non-null int64
retweet_count            2335 non-null int64
retweet_status_id_api    2335 non-null object
dtypes: int64(3), object(4)
memory usage: 127.8+ KB


In [8]:
all_columns = pd.Series(list(df_archive) + list(df_class) + list(df_status))
all_columns[all_columns.duplicated()]

17    tweet_id
29    tweet_id
dtype: object

#### Good that names of columns are not duplicated apart from tweet_id on which the tables can be joined;  but there are actually duplicating contents in several columns (eg text in `df_archive` and text_check in `df_status` )

### Tidying/Cleaning

  - joining `df_archive` and `df_status`, which can sit in one table (following cleaning), complying with tidyness protocol; also once they do, it will be easier to examine the following
  - creating `class_df` from `df_class`
  - dropping rows that have no *favourite_rating* and *retween_count* (this is because those tweets were deleted, and where  subsequntly not received from Twitter API
  - checking tweet text is correct for each tweet (comparing two columns that store tweet text, deciding which to keep),
  - similarly checking other potentially duplicated columns, including tweet_url (columns `expanded_urls` and `tweet_web_url`) and retweet_id (columns `retweeted_status_id` and `retween_status_id_api`)
  - converting the `retweeted_status_id` column to integer (np.int64), in line with tweet_id
  - checking ratings (`rating_numerator` and `rating_denominator`) are correctly imported to ratings columns (from tweet text)
  - creating one column for 'dog_stages' (from the four "dummy" columns of `doggo`, `floofer`, `pupper`, `puppo`)
  - checking 'dog stages' are all correctly extracted
  - convert timestamp from string to datetime
  - identifying and dropping rows that are retweets (explicitly not required to be included in the project)
  - checking other cleanliness issues (e.g. ` dog_picture_url` in df_tweet should be the same as `jpg_url` in df_class; check and drop one)
  - renaming columns in `class_clean`
  
It seems that the image classification table `class_clean` (previously `df_class`) could also be joined without causing tidyness issues, that said, since it is such a different "unit" of information, with its own independent source (ie classification model output), will keep it separate for now.

In [9]:
# create TWEET_CLEAN from df_archive and df_status
tweet_clean = pd.merge(df_archive, df_status,
                            on=['tweet_id'], how='left')

#create CLASS_CLEAN from df_class
class_clean = df_class

In [10]:
tweet_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [11]:
tweet_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 23 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
text_check                    23

#### Dropping rows that have no favourite_rating and retween_count (this is because those tweets were deleted, and where subsequntly not received from Twitter API

In [12]:
tweet_clean = tweet_clean[tweet_clean.favourite_count.notnull()]

In [13]:
tweet_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2335 entries, 0 to 2355
Data columns (total 23 columns):
tweet_id                      2335 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2335 non-null object
source                        2335 non-null object
text                          2335 non-null object
retweeted_status_id           165 non-null float64
retweeted_status_user_id      165 non-null float64
retweeted_status_timestamp    165 non-null object
expanded_urls                 2276 non-null object
rating_numerator              2335 non-null int64
rating_denominator            2335 non-null int64
name                          2335 non-null object
doggo                         2335 non-null object
floofer                       2335 non-null object
pupper                        2335 non-null object
puppo                         2335 non-null object
text_check                    23

In [14]:
tweet_clean.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
667   790337589677002753                    NaN                  NaN   
739   780601303617732608                    NaN                  NaN   
1371  702276748847800320                    NaN                  NaN   
2190  668960084974809088                    NaN                  NaN   
592   798933969379225600                    NaN                  NaN   

                      timestamp  \
667   2016-10-23 23:42:19 +0000   
739   2016-09-27 02:53:48 +0000   
1371  2016-02-23 23:39:59 +0000   
2190  2015-11-24 01:11:27 +0000   
592   2016-11-16 17:01:16 +0000   

                                                 source  \
667   <a href="http://twitter.com/download/iphone" r...   
739   <a href="http://twitter.com/download/iphone" r...   
1371  <a href="http://twitter.com/download/iphone" r...   
2190  <a href="http://twitter.com/download/iphone" r...   
592   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
667   Meet Maggie. She can hear your cells divide. 1...                  NaN   
739   Meet Hercules. He can have whatever he wants f...                  NaN   
1371  "AND IIIIIIIIIIIEIIIIIIIIIIIII WILL ALWAYS LOV...                  NaN   
2190  Meet Jaycob. He got scared of the vacuum. Hide...                  NaN   
592   This is Iroh. He's in a predicament. 12/10 som...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
667                        NaN                        NaN   
739                        NaN                        NaN   
1371                       NaN                        NaN   
2190                       NaN                        NaN   
592                        NaN                        NaN   

                                          expanded_urls  ...  doggo  floofer  \
667   https://twitter.com/dog_rates/status/790337589...  ...   None     None   
739   https://twitter.com/dog_rates/status/780601303...  ...   None     None   
1371  https://twitter.com/dog_rates/status/702276748...  ...   None     None   
2190  https://twitter.com/dog_rates/status/668960084...  ...   None     None   
592   https://twitter.com/dog_rates/status/798933969...  ...   None     None   

     pupper puppo                                         text_check  \
667    None  None  Meet Maggie. She can hear your cells divide. 1...   
739    None  None  Meet Hercules. He can have whatever he wants f...   
1371   None  None  "AND IIIIIIIIIIIEIIIIIIIIIIIII WILL ALWAYS LOV...   
2190   None  None  Meet Jaycob. He got scared of the vacuum. Hide...   
592    None  None  This is Iroh. He's in a predicament. 12/10 som...   

                                     dog_picture_url  \
667   http://pbs.twimg.com/media/CvfX2AnWYAAQTay.jpg   
739   http://pbs.twimg.com/media/CtVAvX-WIAAcGTf.jpg   
1371  http://pbs.twimg.com/media/Cb78-nOWIAENNRc.jpg   
2190  http://pbs.twimg.com/media/CUifpn4WUAAS5X3.jpg   
592   http://pbs.twimg.com/media/CxZiLcLXUAApMVy.jpg   

                                          tweet_web_url favourite_count  \
667   https://twitter.com/dog_rates/status/790337589...          8241.0   
739   https://twitter.com/dog_rates/status/780601303...         12700.0   
1371  https://twitter.com/dog_rates/status/702276748...          2551.0   
2190  https://twitter.com/dog_rates/status/668960084...           702.0   
592   https://twitter.com/dog_rates/status/798933969...         13843.0   

     retweet_count retweet_status_id_api  
667         2005.0                  None  
739         3568.0                  None  
1371         797.0                  None  
2190         240.0                  None  
592         4817.0                  None  

[5 rows x 23 columns]

In [15]:
#few checks to see values are sensible
tweet_clean.favourite_count.sort_values().head(5)

1295    51.0
342     68.0
2339    75.0
2298    82.0
2319    99.0
Name: favourite_count, dtype: float64

In [16]:
tweet_clean.favourite_count.sort_values().tail(5)

1079    119452.0
135     120097.0
534     124911.0
413     137875.0
1039    161664.0
Name: favourite_count, dtype: float64

In [17]:
tweet_clean.retweet_count.sort_values().head(5)

291     1.0
342     2.0
1295    2.0
274     2.0
113     3.0
Name: retweet_count, dtype: float64

In [18]:
tweet_clean.retweet_count.sort_values().tail(5)

66      42651.0
413     46757.0
534     59854.0
1079    60745.0
1039    82098.0
Name: retweet_count, dtype: float64

#### Checking tweet text is correct for each tweet (comparing two columns that store tweet text, deciding which to keep)

In [19]:
sum(tweet_clean.text != tweet_clean.text_check)

0

Ok, marvelous, all text from tweets was downloaded correctly;
therefore can just drop one of the text columns, will drop `text_check`

In [20]:
tweet_clean.drop(columns=['text_check'], axis=1, inplace=True)

C:\Users\vilij\Anaconda3\lib\site-packages\pandas\core\frame.py:3940: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [21]:
tweet_clean.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  ...     name  doggo  \
0  https://twitter.com/dog_rates/status/892420643...  ...  Phineas   None   
1  https://twitter.com/dog_rates/status/892177421...  ...    Tilly   None   

  floofer pupper puppo                                 dog_picture_url  \
0    None   None  None  http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   
1    None   None  None  http://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   

                                       tweet_web_url favourite_count  \
0  https://twitter.com/dog_rates/status/892420643...         37434.0   
1  https://twitter.com/dog_rates/status/892177421...         32164.0   

  retweet_count  retweet_status_id_api  
0        8138.0                   None  
1        6029.0                   None  

[2 rows x 22 columns]

#### Similarly checking other potentially duplicated columns, including tweet_url (columns `expanded_urls` and `tweet_web_url`) and retweet_id (columns `retweeted_status_id` and  `retween_status_id_api`)

In [22]:
sum(tweet_clean.expanded_urls != tweet_clean.tweet_web_url)

852

Hm, not such a great match..

In [23]:
tweet_clean[tweet_clean.expanded_urls != tweet_clean.tweet_web_url][['expanded_urls','tweet_web_url']]

expanded_urls  \
4     https://twitter.com/dog_rates/status/891327558...   
6     https://gofundme.com/ydvmve-surgery-for-jax,ht...   
7     https://twitter.com/dog_rates/status/890729181...   
10    https://twitter.com/dog_rates/status/890006608...   
13    https://twitter.com/dog_rates/status/889638837...   
17    https://twitter.com/dog_rates/status/888804989...   
18    https://twitter.com/dog_rates/status/888554962...   
20    https://twitter.com/dog_rates/status/888078434...   
23    https://twitter.com/dog_rates/status/887473957...   
26    https://twitter.com/dog_rates/status/886983233...   
27    https://www.gofundme.com/mingusneedsus,https:/...   
29    https://twitter.com/dog_rates/status/886366144...   
30                                                  NaN   
32    https://twitter.com/dog_rates/status/886053434...   
35    https://twitter.com/4bonds2carbon/status/88551...   
36    https://twitter.com/dog_rates/status/830583320...   
37    https://twitter.com/dog_rates/status/885167619...   
39    https://twitter.com/dog_rates/status/884876753...   
42    https://twitter.com/kaijohnson_19/status/88396...   
45    https://twitter.com/dog_rates/status/883482846...   
47    https://twitter.com/dog_rates/status/883117836...   
48    https://twitter.com/dog_rates/status/882992080...   
51    https://twitter.com/dog_rates/status/882268110...   
52    https://twitter.com/dog_rates/status/882045870...   
55                                                  NaN   
60    https://twitter.com/dog_rates/status/880465832...   
61    https://twitter.com/dog_rates/status/880221127...   
63    https://twitter.com/dog_rates/status/879862464...   
64                                                  NaN   
68    https://twitter.com/dog_rates/status/878057613...   
...                                                 ...   
1906  https://twitter.com/dog_rates/status/674468880...   
1914                                                NaN   
1916                      https://vine.co/v/i7nWzrenw5h   
1939  https://twitter.com/dog_rates/status/673887867...   
1940                                                NaN   
1945  https://twitter.com/dog_rates/status/673707060...   
1948  https://twitter.com/dog_rates/status/673697980...   
1949  https://twitter.com/dog_rates/status/673689733...   
1962  https://twitter.com/dog_rates/status/673355879...   
1968  https://twitter.com/dog_rates/status/673320132...   
1969  https://twitter.com/dog_rates/status/673317986...   
2005  https://twitter.com/dog_rates/status/672272411...   
2027  https://twitter.com/dog_rates/status/671874878...   
2031  https://twitter.com/dog_rates/status/671768281...   
2035  https://twitter.com/dog_rates/status/671735591...   
2038                                                NaN   
2039  https://twitter.com/dog_rates/status/671547767...   
2049  https://twitter.com/dog_rates/status/671504605...   
2056  https://twitter.com/dog_rates/status/671357843...   
2058  https://twitter.com/dog_rates/status/671347597...   
2073  https://twitter.com/dog_rates/status/670995969...   
2084  https://twitter.com/dog_rates/status/670807719...   
2149                                                NaN   
2189                                                NaN   
2205  https://twitter.com/dog_rates/status/668633411...   
2209  https://twitter.com/dog_rates/status/668623201...   
2212                      https://vine.co/v/ea0OwvPTx9l   
2259  https://twitter.com/dogratingrating/status/667...   
2260  https://twitter.com/dogratingrating/status/667...   
2298                                                NaN   

                                          tweet_web_url  
4     https://twitter.com/dog_rates/status/891327558...  
6     https://twitter.com/dog_rates/status/890971913...  
7     https://twitter.com/dog_rates/status/890729181...  
10    https://twitter.com/dog_rates/status/890006608...  
13    https://twitter.com/dog_rates/status/889638837...  
17    https://twitter.com/dog_rates

In [24]:
tweet_clean[tweet_clean.expanded_urls != tweet_clean.tweet_web_url].loc[2209,['expanded_urls','tweet_web_url']].values

array(['https://twitter.com/dog_rates/status/668623201287675904/photo/1,https://twitter.com/dog_rates/status/668623201287675904/photo/1,https://twitter.com/dog_rates/status/668623201287675904/photo/1,https://twitter.com/dog_rates/status/668623201287675904/photo/1',
       'https://twitter.com/dog_rates/status/668623201287675904/photo/1'],
      dtype=object)

Ok, seems that `tweet_web_url` is more complete and also devoid of repetitions/multiple entries (unlike `expanded_urls` column); threrefore will drop `expanded_urls` column

In [25]:
tweet_clean.drop(columns=['expanded_urls'], axis=1, inplace=True)

C:\Users\vilij\Anaconda3\lib\site-packages\pandas\core\frame.py:3940: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [26]:
tweet_clean.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  rating_numerator  ...  \
0                       NaN                        NaN                13  ...   
1                       NaN                        NaN                13  ...   

      name doggo floofer pupper puppo  \
0  Phineas  None    None   None  None   
1    Tilly  None    None   None  None   

                                  dog_picture_url  \
0  http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   
1  http://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   

                                       tweet_web_url favourite_count  \
0  https://twitter.com/dog_rates/status/892420643...         37434.0   
1  https://twitter.com/dog_rates/status/892177421...         32164.0   

   retweet_count  retweet_status_id_api  
0         8138.0                   None  
1         6029.0                   None  

[2 rows x 21 columns]

#### Now checking for duplications in retweet_id (columns `retweeted_status_id` and  `retween_status_id_api`)

In [27]:
#normalising for NaN and 'None'
tweet_clean.retweet_status_id_api[tweet_clean.retweet_status_id_api == 'None'] = np.nan

C:\Users\vilij\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\vilij\Anaconda3\lib\site-packages\pandas\core\generic.py:8682: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)
C:\Users\vilij\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2910: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


In [28]:
sum(tweet_clean.retweet_status_id_api == 'None')

0

In [29]:
sum(tweet_clean.retweet_status_id_api.isnull())

2170

In [30]:
#Ok, now let's compare the two columns
tweet_clean[tweet_clean.retweeted_status_id != tweet_clean.retweet_status_id_api][['retweeted_status_id','retweet_status_id_api']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2335 entries, 0 to 2355
Data columns (total 2 columns):
retweeted_status_id      165 non-null float64
retweet_status_id_api    165 non-null object
dtypes: float64(1), object(1)
memory usage: 54.7+ KB


In [31]:
#ok, for comparison let's try to use `retweet_status_id_api` values as float (also NaNs are accomodated for that way)
sum(tweet_clean.retweeted_status_id == tweet_clean.retweet_status_id_api.astype(float))

165

In [32]:
temp = tweet_clean.retweet_status_id_api.astype(float)
temp2 = tweet_clean.retweeted_status_id
sum(temp[temp.notnull()] != temp2[temp2.notnull()])

0

#### Ok, so in conclusion, the `retweeted_status_id`  and `retweet_status_id_api` are the same; therefore I will drop the `retweet_status_id_api` column

In [33]:
tweet_clean.drop(columns=['retweet_status_id_api'], axis=1, inplace=True)

C:\Users\vilij\Anaconda3\lib\site-packages\pandas\core\frame.py:3940: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [34]:
tweet_clean.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  rating_numerator  \
0                       NaN                        NaN                13   
1                       NaN                        NaN                13   
2                       NaN                        NaN                12   

   rating_denominator     name doggo floofer pupper puppo  \
0                  10  Phineas  None    None   None  None   
1                  10    Tilly  None    None   None  None   
2                  10   Archie  None    None   None  None   

                                  dog_picture_url  \
0  http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   
1  http://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2  http://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   

                                       tweet_web_url  favourite_count  \
0  https://twitter.com/dog_rates/status/892420643...          37434.0   
1  https://twitter.com/dog_rates/status/892177421...          32164.0   
2  https://twitter.com/dog_rates/status/891815181...          24251.0   

   retweet_count  
0         8138.0  
1         6029.0  
2         3991.0

#### For completeness with the `retweeted_status_id` column, will also convert it to integer (np.int64), in line with tweet_id
To deal with NaNs I will first convert all NaNs to 0s, then the column to int64

In [35]:
tweet_clean.retweeted_status_id[tweet_clean.retweeted_status_id.isnull()] = 0
tweet_clean.retweeted_status_id = tweet_clean.retweeted_status_id.astype(np.int64)
#tweet_clean.retweeted_status_id[tweet_clean.retweeted_status_id == 0] = np.nan

C:\Users\vilij\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\vilij\Anaconda3\lib\site-packages\pandas\core\generic.py:8682: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)
C:\Users\vilij\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2910: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  exec(code_obj, self.user_global_ns, self.user_ns)
C

In [36]:
tweet_clean.retweeted_status_id.dtype

dtype('int64')

#### Checking ratings are correctly imported to ratings columns (from tweet text)

In [37]:
tweet_clean.loc[0].text

"This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU"

In [38]:
test = tweet_clean.text.str.extract(r'(\d{1,4}[\.]?[\d{1,2}]?)([\s]?[/][\s]?)(\d{2,4})',expand = True)
test

0  1   2
0     13  /  10
1     13  /  10
2     12  /  10
3     13  /  10
4     12  /  10
5     13  /  10
6     13  /  10
7     13  /  10
8     13  /  10
9     14  /  10
10    13  /  10
11    13  /  10
12    13  /  10
13    12  /  10
14    13  /  10
15    13  /  10
16    12  /  10
17    13  /  10
18    13  /  10
20    12  /  10
21    13  /  10
22    14  /  10
23    13  /  10
24    13  /  10
25    12  /  10
26    13  /  10
27    13  /  10
28    13  /  10
29    12  /  10
30    12  /  10
...   .. ..  ..
2326   2  /  10
2327   7  /  10
2328   9  /  10
2329  11  /  10
2330   6  /  10
2331   8  /  10
2332  10  /  10
2333   9  /  10
2334   3  /  10
2335   9  /  10
2336  11  /  10
2337  10  /  10
2338   1  /  10
2339  11  /  10
2340   8  /  10
2341   9  /  10
2342   6  /  10
2343  10  /  10
2344   9  /  10
2345  10  /  10
2346   8  /  10
2347   9  /  10
2348  10  /  10
2349   2  /  10
2350  10  /  10
2351   5  /  10
2352   6  /  10
2353   9  /  10
2354   7  /  10
2355   8  /  10

[2335 rows x 3 columns]

In [39]:
sum(test[0].isnull())

1

In [40]:
#let's check where NaNs are coming from
test = test.join(tweet_clean['text'], how='left')
test.to_csv('test.csv')

In [41]:
tweet_clean.rating_numerator.value_counts()

12      554
11      464
10      459
13      340
9       157
8       102
7        55
14       51
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [42]:
test[0].value_counts()

12      554
11      464
10      459
13      340
9       158
8       102
7        54
14       51
5        35
6        32
3        19
4        17
2         9
1         8
75        2
420       2
15        2
0         2
204       1
007       1
960       1
143       1
84        1
27        1
13.5      1
165       1
80        1
99        1
60        1
1776      1
121       1
45        1
9.5       1
666       1
44        1
26        1
182       1
50        1
88        1
20        1
144       1
17        1
Name: 0, dtype: int64

In [43]:
tweet_clean.rating_denominator.value_counts()

10     2313
50        3
80        2
11        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [44]:
test[2].value_counts()

10     2314
50        3
80        2
20        2
11        2
16        1
00        1
170       1
120       1
150       1
90        1
15        1
70        1
130       1
40        1
110       1
Name: 2, dtype: int64

#### Overall quality of extratction looks good (check in CSV for exceptions, because it's much easier to see full text there); 
It seems that the really high values in denominator and occosationally numerator eg 1776 or 960 are actually valid ratings, and may be in some cases ratings of multiple of dogs in one picture.  Evcen 0 in denominator is a valid number, which was given twice, what it seems not to dogs but to random picture sent to @dog_rates. One real exception was that code reads 9.75 as 75.
For now, will not delete any extreme values, because they're valid, will deal with these in visuals/analysis as relevant.

#### Therefore will now attach new ratings to `tweet_clean` dataframe, to carry out comparison between existings ratings there and newly extracted ratings, and to drop or keep columns accordingly.

In [45]:
test.columns

Index([0, 1, 2, 'text'], dtype='object')

In [46]:
tweet_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'rating_numerator', 'rating_denominator',
       'name', 'doggo', 'floofer', 'pupper', 'puppo', 'dog_picture_url',
       'tweet_web_url', 'favourite_count', 'retweet_count'],
      dtype='object')

In [47]:
#drop uncenessary columns from test dataframe
test.drop(columns=[1,'text'], axis = 1, inplace=True)

#rename columns
test.rename(columns={0:'rating_numerator_alt',2:'rating_denominator_alt'}, inplace=True)

#join test to tweet_clean
tweet_clean = tweet_clean.join(test, how='left')

In [48]:
tweet_clean.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                    0   
1  This is Tilly. She's just checking pup on you....                    0   

   retweeted_status_user_id retweeted_status_timestamp  rating_numerator  ...  \
0                       NaN                        NaN                13  ...   
1                       NaN                        NaN                13  ...   

   doggo floofer pupper puppo                                 dog_picture_url  \
0   None    None   None  None  http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   
1   None    None   None  None  http://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   

                                       tweet_web_url favourite_count  \
0  https://twitter.com/dog_rates/status/892420643...         37434.0   
1  https://twitter.com/dog_rates/status/892177421...         32164.0   

  retweet_count  rating_numerator_alt  rating_denominator_alt  
0        8138.0                    13                      10  
1        6029.0                    13                      10  

[2 rows x 22 columns]

In [49]:
tweet_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2335 entries, 0 to 2355
Data columns (total 22 columns):
tweet_id                      2335 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2335 non-null object
source                        2335 non-null object
text                          2335 non-null object
retweeted_status_id           2335 non-null int64
retweeted_status_user_id      165 non-null float64
retweeted_status_timestamp    165 non-null object
rating_numerator              2335 non-null int64
rating_denominator            2335 non-null int64
name                          2335 non-null object
doggo                         2335 non-null object
floofer                       2335 non-null object
pupper                        2335 non-null object
puppo                         2335 non-null object
dog_picture_url               2335 non-null object
tweet_web_url                 233

#### In order to do the comparison between existing ratings numerator and denominator columns (`rating_numerator` and `rating_denominator`) and newly extracted ones, we will need to convert both to float -- there were some ratings (e.g. 13.5/10) that were not intergers, so I decided to follow this rule

In [50]:
sum(tweet_clean.rating_numerator.astype(float) != tweet_clean.rating_numerator_alt.astype(float))

4

In [51]:
print(tweet_clean[tweet_clean.rating_numerator.astype(float) != tweet_clean.rating_numerator_alt.astype(float)]['text'].values)

['This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948'
 'Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx'
 "I've been told there's a slight possibility he's checking his mirror. We'll bump to 9.5/10. Still a menace"
 'This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv']


In [52]:
tweet_clean[tweet_clean.rating_numerator.astype(float) != tweet_clean.rating_numerator_alt.astype(float)][['rating_numerator','rating_numerator_alt']]

rating_numerator rating_numerator_alt
45                   5                 13.5
516                 24                  NaN
1689                 5                  9.5
2335                 1                    9

#### Conclusion:
Newly extraceted ratings numerators (in column `rating_numerator_alt`) are the accurate ones for these 4 instances of mismatch

In [53]:
sum(tweet_clean.rating_denominator.astype(float) != tweet_clean.rating_denominator_alt.astype(float))

2

In [54]:
print(tweet_clean[tweet_clean.rating_denominator.astype(float) != tweet_clean.rating_denominator_alt.astype(float)]['text'].values)

['Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx'
 'This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv']


In [55]:
tweet_clean[tweet_clean.rating_denominator.astype(float) != tweet_clean.rating_denominator_alt.astype(float)][['rating_denominator','rating_denominator_alt']]

rating_denominator rating_denominator_alt
516                    7                    NaN
2335                   2                     10

#### Conclusion:
Newly extraceted ratings denominators (in column `rating_denominator_alt`) are the accurate ones for these 2 instances of mismatch

#### ACTION:
Drop existing ratings columns and replace them with newly extracted ratings columns.

In [56]:
tweet_clean.drop(columns=['rating_numerator','rating_denominator'], axis=1, inplace=True)
tweet_clean.rename(columns={'rating_numerator_alt':'rating_numerator','rating_denominator_alt':'rating_denominator'}, inplace=True)

In [57]:
tweet_clean.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                    0   
1  This is Tilly. She's just checking pup on you....                    0   

   retweeted_status_user_id retweeted_status_timestamp     name doggo floofer  \
0                       NaN                        NaN  Phineas  None    None   
1                       NaN                        NaN    Tilly  None    None   

  pupper puppo                                 dog_picture_url  \
0   None  None  http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   
1   None  None  http://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   

                                       tweet_web_url  favourite_count  \
0  https://twitter.com/dog_rates/status/892420643...          37434.0   
1  https://twitter.com/dog_rates/status/892177421...          32164.0   

   retweet_count rating_numerator rating_denominator  
0         8138.0               13                 10  
1         6029.0               13                 10

#### Will also fix the two numerator values that were 75 instead of actual 9.5, and convert all values to Float type, to complete cleaning of this column 

In [58]:
print(tweet_clean[tweet_clean.rating_numerator.astype(float) == 75][['text','rating_numerator']].values)

[["RT @dog_rates: This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wu…"
  '75']
 ["This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS"
  '75']]


In [59]:
tweet_clean.loc[tweet_clean.rating_numerator.astype(float) == 75,'rating_numerator'] = 9.75

In [60]:
tweet_clean.loc[tweet_clean.rating_numerator.astype(float) == 75,'rating_numerator']

Series([], Name: rating_numerator, dtype: object)

In [61]:
tweet_clean.rating_numerator = tweet_clean.rating_numerator.astype(float)

In [62]:
tweet_clean.rating_denominator = tweet_clean.rating_denominator.astype(float)

#### Will now proceed to check if 'dog stages' were correctly extracted, steps will include:
* Creating one column for 'dog_stages' (from the four "dummy" columns of `doggo`, `floofer`, `pupper`, `puppo`)
* Extracting 'dog stages' anew
* Comparing to existing 'dog stages'
* Identifying and keeping accurate column, dropping inaccurate column

In [63]:
tweet_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'name', 'doggo', 'floofer', 'pupper',
       'puppo', 'dog_picture_url', 'tweet_web_url', 'favourite_count',
       'retweet_count', 'rating_numerator', 'rating_denominator'],
      dtype='object')

In [64]:
tweet_clean.loc[tweet_clean.puppo == 'None','puppo'] = ''
tweet_clean.loc[tweet_clean.doggo == 'None','doggo'] = ''
tweet_clean.loc[tweet_clean.floofer == 'None','floofer'] = ''
tweet_clean.loc[tweet_clean.pupper == 'None','pupper'] = ''

In [65]:
#checking the possible outcome of concatenation
(tweet_clean.puppo+tweet_clean.doggo+tweet_clean.floofer+tweet_clean.pupper).value_counts()

                1959
pupper           244
doggo             80
puppo             29
doggopupper       12
floofer            9
doggofloofer       1
puppodoggo         1
dtype: int64

Ok, looks like the given extraction was not too accurate, but the concatenation did work, so proceeding..

In [66]:
tweet_clean['dog_stages'] = tweet_clean.puppo+tweet_clean.doggo+tweet_clean.floofer+tweet_clean.pupper

In [67]:
#Quick sample check
tweet_clean.sample().values

array([[667065535570550784, nan, nan, '2015-11-18 19:43:11 +0000',
        '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
        'Here we have a Hufflepuff. Loves vest. Eyes wide af. Flaccid tail. Matches carpet. Always a little blurry. 8/10 https://t.co/7JdgVqDnvR',
        0, nan, nan, 'None', '', '', '', '',
        'http://pbs.twimg.com/media/CUHkkJpXIAA2w3n.jpg',
        'https://twitter.com/dog_rates/status/667065535570550784/photo/1',
        163.0, 46.0, 8.0, 10.0, '']], dtype=object)

In [68]:
#Checking out "floofer" class which is actually not defined, but will include it as separate from 'floof' and see what happens
tweet_clean[tweet_clean.floofer != ''].sample().values

array([[800388270626521089, nan, nan, '2016-11-20 17:20:08 +0000',
        '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
        'This is Doc. He takes time out of every day to worship our plant overlords. 12/10 quite the floofer https://t.co/azMneS6Ly5',
        0, nan, nan, 'Doc', '', 'floofer', '', '',
        'http://pbs.twimg.com/media/CxuM3ofWgAE72kK.jpg',
        'https://twitter.com/dog_rates/status/800388270626521089/photo/1',
        11758.0, 3025.0, 12.0, 10.0, 'floofer']], dtype=object)

In [69]:
#extract anew
test = tweet_clean.text.str.extractall(r'([Pp]uppo|[Dd]oggo|[Pp]upper|[Bb]lep|[Ss]noot|[Ff]loofer|[Ff]loof)').unstack()
test[0][0].value_counts()

pupper     262
doggo       88
puppo       37
floof       20
Floof       10
Doggo        8
Pupper       8
Floofer      5
floofer      3
Blep         2
Puppo        1
blep         1
Name: 0, dtype: int64

In [70]:
test = test[0]

In [71]:
#attach new values, lowercase them
test[0] = test[0].str.lower()
test[1] = test[1].str.lower()
test[2] = test[2].str.lower()
test[3] = test[3].str.lower()
test.sample(5)

match       0    1    2    3
351     doggo  NaN  NaN  NaN
331     doggo  NaN  NaN  NaN
1770   pupper  NaN  NaN  NaN
1814   pupper  NaN  NaN  NaN
1220   pupper  NaN  NaN  NaN

In [72]:
test.rename(columns={0:'dog_stages_alt1', 1:'dog_stages_alt2', 2:'dog_stages_alt3', 3:'dog_stages_alt4'}, inplace=True)
tweet_clean = tweet_clean.join(test, how='left')

In [73]:
#Quick sample check
tweet_clean[(tweet_clean.dog_stages_alt1.notnull())&(tweet_clean.dog_stages_alt2.notnull())]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
29    886366144734445568                    NaN                  NaN   
46    883360690899218434                    NaN                  NaN   
172   858843525470990336                    NaN                  NaN   
191   855851453814013952                    NaN                  NaN   
200   854010172552949760                    NaN                  NaN   
433   820690176645140481                    NaN                  NaN   
460   817777686764523521                    NaN                  NaN   
531   808106460588765185                    NaN                  NaN   
565   802265048156610565           7.331095e+17         4.196984e+09   
575   801115127852503040                    NaN                  NaN   
631   794205286408003585                    NaN                  NaN   
705   785639753186217984                    NaN                  NaN   
733   781308096455073793                    NaN                  NaN   
764   777953400541634568                    NaN                  NaN   
778   775898661951791106                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
835   768193404517830656                    NaN                  NaN   
881   760521673607086080                    NaN                  NaN   
889   759793422261743616                    NaN                  NaN   
956   751583847268179968                    NaN                  NaN   
1063  741067306818797568                    NaN                  NaN   
1113  733109485275860992                    NaN                  NaN   
1152  725786712245440512                    NaN                  NaN   
1304  707411934438625280                    NaN                  NaN   
1340  704859558691414016                    NaN                  NaN   
1367  702598099714314240                    NaN                  NaN   
1382  700864154249383937                    NaN                  NaN   
1653  683462770029932544                    NaN                  NaN   
1788  677557565589463040                    NaN                  NaN   
1828  676263575653122048                    NaN                  NaN   
1907  674447403907457024                    NaN                  NaN   

                      timestamp  \
29    2017-07-15 23:25:31 +0000   
46    2017-07-07 16:22:55 +0000   
172   2017-05-01 00:40:27 +0000   
191   2017-04-22 18:31:02 +0000   
200   2017-04-17 16:34:26 +0000   
433   2017-01-15 17:52:40 +0000   
460   2017-01-07 16:59:28 +0000   
531   2016-12-12 00:29:28 +0000   
565   2016-11-25 21:37:47 +0000   
575   2016-11-22 17:28:25 +0000   
631   2016-11-03 15:51:10 +0000   
705   2016-10-11 00:34:48 +0000   
733   2016-09-29 01:42:20 +0000   
764   2016-09-19 19:31:59 +0000   
778   2016-09-14 03:27:11 +0000   
822   2016-08-29 03:00:36 +0000   
835   2016-08-23 21:09:14 +0000   
881   2016-08-02 17:04:31 +0000   
889   2016-07-31 16:50:42 +0000   
956   2016-07-09 01:08:47 +0000   
1063  2016-06-10 00:39:48 +0000   
1113  2016-05-19 01:38:16 +0000   
1152  2016-04-28 20:40:11 +0000   
1304  2016-03-09 03:45:22 +0000   
1340  2016-03-02 02:43:09 +0000   
1367  2016-02-24 20:56:55 +0000   
1382  2016-02-20 02:06:50 +0000   
1653  2016-01-03 01:39:57 +0000   
1788  2015-12-17 18:34:46 +0000   
1828  2015-12-14 04:52:55 +0000   
1907  2015-12-09 04:36:06 +0000   

                                                 source  \
29    <a href="http://twitter.com/download/iphone" r...   
46    <a href="http://twitter.com/download/iphone" r...   
172   <a href="http://twitter.com/download/iphone" r...   
191   <a href="http://twitter.com/download/iphone" r...   
200   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
460   <a href="http://twitter.com/download/iphone" r...   
531   <a href="http://twitter.com/download/iphone" r...   
565   <a href="http://twitter.com/download/iphone" r...   


In [74]:
tweet_clean.loc[822].values

array([770093767776997377, nan, nan, '2016-08-29 03:00:36 +0000',
       '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       'RT @dog_rates: This is just downright precious af. 12/10 for both pupper and doggo https://t.co/o5J479bZUC',
       741067306818797568, 4196983835.0, '2016-06-10 00:39:48 +0000',
       'just', 'doggo', '', 'pupper', '',
       'http://pbs.twimg.com/media/CkjMx99UoAM2B1a.jpg',
       'https://twitter.com/dog_rates/status/741067306818797568/photo/1',
       9761.0, 3238.0, 12.0, 10.0, 'doggopupper', 'pupper', 'doggo', nan,
       nan], dtype=object)

In [75]:
tweet_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'name', 'doggo', 'floofer', 'pupper',
       'puppo', 'dog_picture_url', 'tweet_web_url', 'favourite_count',
       'retweet_count', 'rating_numerator', 'rating_denominator', 'dog_stages',
       'dog_stages_alt1', 'dog_stages_alt2', 'dog_stages_alt3',
       'dog_stages_alt4'],
      dtype='object')

#### Quick examination shows that the new extraction is more complete than existing one (e.g. picks up on floof, blep; even though its not black and white if there is a reference to floof that the dog stage is floof, sometimes the reference is purely to floofy nature of the dog)
#### Each tweet may contain more than one dog and therefore contain doggo and pupper, or one dog within single tweet can have several stages attributed to him (sometimes inadvertently, through an adjactive that contains dog stage indicator);
#### To really tidy it up one should create a separate table, that should contain tweet_id and dog_stage matches;
#### Also, a serious consideration should be given if floofer and floof is indeed one category;
#### But for here, will simply drop all the columns associated with existing classification ( `doggo`, `floofer`, `pupper`, `puppo` and  `dog_stages`) and leave four columns wiht newly extracted categories (that is `dog_stages_alt1`, `dog_stages_alt2`,`dog_stages_alt3`,`dog_stages_alt4`)

In [76]:
tweet_clean.drop(columns=['doggo','floofer','pupper','puppo','dog_stages'], axis=1, inplace=True)

In [77]:
tweet_clean.sample(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
58   880935762899988482                    NaN                  NaN   
718  783466772167098368                    NaN                  NaN   

                     timestamp  \
58   2017-06-30 23:47:07 +0000   
718  2016-10-05 00:40:09 +0000   

                                                source  \
58   <a href="http://twitter.com/download/iphone" r...   
718  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
58   This is Louis. He's crossing. It's a big deal....                    0   
718  This is Finn. He likes eavesdropping from fili...                    0   

     retweeted_status_user_id retweeted_status_timestamp   name  \
58                        NaN                        NaN  Louis   
718                       NaN                        NaN   Finn   

                                    dog_picture_url  \
58   http://pbs.twimg.com/media/DDm2Z5aXUAEDS2u.jpg   
718  http://pbs.twimg.com/media/Ct9u3ljW8AEnVIm.jpg   

                                         tweet_web_url  favourite_count  \
58   https://twitter.com/dog_rates/status/880935762...          16498.0   
718  https://twitter.com/dog_rates/status/783466772...           8879.0   

     retweet_count  rating_numerator  rating_denominator dog_stages_alt1  \
58          2686.0              13.0                10.0             NaN   
718         2408.0              11.0                10.0             NaN   

    dog_stages_alt2 dog_stages_alt3 dog_stages_alt4  
58              NaN             NaN             NaN  
718             NaN             NaN             NaN

#### timestamp from string to datetime

In [78]:
# To datetime
tweet_clean.timestamp = pd.to_datetime(tweet_clean.timestamp)

In [79]:
tweet_clean.timestamp.dtype

datetime64[ns, UTC]

In [80]:
tweet_clean.sample(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
785  775085132600442880                    NaN                  NaN   
35   885518971528720385                    NaN                  NaN   

                    timestamp  \
785 2016-09-11 21:34:30+00:00   
35  2017-07-13 15:19:09+00:00   

                                                source  \
785  <a href="http://twitter.com/download/iphone" r...   
35   <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
785  This is Tucker. He would like a hug. 13/10 som...                    0   
35   I have a new hero and his name is Howard. 14/1...                    0   

     retweeted_status_user_id retweeted_status_timestamp    name  \
785                       NaN                        NaN  Tucker   
35                        NaN                        NaN    None   

                                    dog_picture_url  \
785  http://pbs.twimg.com/media/CsGnz64WYAEIDHJ.jpg   
35                                             None   

                                         tweet_web_url  favourite_count  \
785  https://twitter.com/dog_rates/status/775085132...          16281.0   
35   https://twitter.com/dog_rates/status/830583320...          19777.0   

     retweet_count  rating_numerator  rating_denominator dog_stages_alt1  \
785         5031.0              13.0                10.0             NaN   
35          3596.0              14.0                10.0             NaN   

    dog_stages_alt2 dog_stages_alt3 dog_stages_alt4  
785             NaN             NaN             NaN  
35              NaN             NaN             NaN

#### Identifying and dropping rows that are retweets (explicitly not required to be included in the project)
* any tweets that have retweeted_status_id as non zero value will be retweets, so will select those rows and drop them
* prior will visually cross check with other retweet attributes such as  retweeted_status_user_id, retweeted_status_timestamp
* will also drop the columns that are retweet attributes, as they would in any case be NaNs

In [81]:
tweet_clean[tweet_clean.retweeted_status_id != 0].sample(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
285  838916489579200512                    NaN                  NaN   
343  832040443403784192                    NaN                  NaN   
767  777641927919427584                    NaN                  NaN   

                    timestamp  \
285 2017-03-07 00:57:32+00:00   
343 2017-02-16 01:34:34+00:00   
767 2016-09-18 22:54:18+00:00   

                                                source  \
285  <a href="http://twitter.com/download/iphone" r...   
343  <a href="http://twitter.com/download/iphone" r...   
767  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
285  RT @KibaDva: I collected all the good dogs!! 1...   838905980628819968   
343  RT @dog_rates: This is Klein. These pics were ...   769940425801170944   
767  RT @dog_rates: This is Arnie. He's a Nova Scot...   750429297815552000   

     retweeted_status_user_id retweeted_status_timestamp   name  \
285              8.117408e+08  2017-03-07 00:15:46 +0000   None   
343              4.196984e+09  2016-08-28 16:51:16 +0000  Klein   
767              4.196984e+09  2016-07-05 20:41:01 +0000  Arnie   

                                    dog_picture_url  \
285  http://pbs.twimg.com/media/C6RkhU6UYAAMDpb.jpg   
343  http://pbs.twimg.com/media/Cq9guJ5WgAADfpF.jpg   
767  http://pbs.twimg.com/media/CmoPdmHW8AAi8BI.jpg   

                                         tweet_web_url  favourite_count  \
285  https://twitter.com/KibaDva/status/83890598062...            738.0   
343  https://twitter.com/dog_rates/status/769940425...          32790.0   
767  https://twitter.com/dog_rates/status/750429297...          13651.0   

     retweet_count  rating_numerator  rating_denominator dog_stages_alt1  \
285           36.0              15.0                10.0             NaN   
343        10240.0              12.0                10.0             NaN   
767         4527.0              12.0                10.0           floof   

    dog_stages_alt2 dog_stages_alt3 dog_stages_alt4  
285             NaN             NaN             NaN  
343             NaN             NaN             NaN  
767             NaN             NaN             NaN

In [82]:
tweet_clean = tweet_clean[tweet_clean.retweeted_status_id == 0]

In [83]:
tweet_clean[tweet_clean.retweeted_status_id != 0].shape

(0, 20)

In [84]:
tweet_clean.drop(columns=['retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'], axis=1, inplace=True)

In [85]:
tweet_clean.sample(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1788  677557565589463040                    NaN                  NaN   
257   843856843873095681                    NaN                  NaN   
2140  670003130994700288                    NaN                  NaN   

                     timestamp  \
1788 2015-12-17 18:34:46+00:00   
257  2017-03-20 16:08:44+00:00   
2140 2015-11-26 22:16:09+00:00   

                                                 source  \
1788  <a href="http://twitter.com/download/iphone" r...   
257   <a href="http://twitter.com/download/iphone" r...   
2140  <a href="http://twitter.com/download/iphone" r...   

                                                   text     name  \
1788  Reckless pupper here. Not even looking at road...     None   
257   Say hello to Sadie and Daisy. They do all thei...    Sadie   
2140  This is Raphael. He is a Baskerville Conquista...  Raphael   

                                     dog_picture_url  \
1788  http://pbs.twimg.com/media/CWcrAVQWEAA6QMp.jpg   
257   http://pbs.twimg.com/media/C7X7Ui0XgAA3m19.jpg   
2140  http://pbs.twimg.com/media/CUxUSuaW4AAdQzv.jpg   

                                          tweet_web_url  favourite_count  \
1788  https://twitter.com/dog_rates/status/677557565...           2490.0   
257   https://twitter.com/dog_rates/status/843856843...          21985.0   
2140  https://twitter.com/dog_rates/status/670003130...            333.0   

      retweet_count  rating_numerator  rating_denominator dog_stages_alt1  \
1788         1204.0              10.0                10.0          pupper   
257          4814.0              12.0                10.0             NaN   
2140           93.0              10.0                10.0             NaN   

     dog_stages_alt2 dog_stages_alt3 dog_stages_alt4  
1788          pupper             NaN             NaN  
257              NaN             NaN             NaN  
2140             NaN             NaN             NaN

In [86]:
tweet_clean.shape

(2170, 17)

#### Check if ` dog_picture_url` in df_tweet should be the same as `jpg_url` in df_class; check and drop one
#### Will check, but ultimatelly, since the algorithm for identifying breeds was run on `jpg_url`, this column is the one to keep (unless one had access to the algorithm and could reidentify the breed)

In [87]:
tweet_clean.shape, class_clean.shape

((2170, 17), (2075, 12))

In [88]:
picture_df = pd.merge(tweet_clean, class_clean,
                            on=['tweet_id'], how='left')

In [89]:
picture_df.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'name', 'dog_picture_url', 'tweet_web_url',
       'favourite_count', 'retweet_count', 'rating_numerator',
       'rating_denominator', 'dog_stages_alt1', 'dog_stages_alt2',
       'dog_stages_alt3', 'dog_stages_alt4', 'jpg_url', 'img_num', 'p1',
       'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf',
       'p3_dog'],
      dtype='object')

In [90]:
picture_df.sample().values

array([[722613351520608256, nan, nan,
        Timestamp('2016-04-20 02:30:23+0000', tz='UTC'),
        '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
        "Meet Schnitzel. He's a Tropicana Floofboop. Getting too big for his favorite basket. 12/10 just so damn fluffy https://t.co/qjd0UJKYUY",
        'Schnitzel', 'http://pbs.twimg.com/media/Cgc9AjMVIAERdUA.jpg',
        'https://twitter.com/dog_rates/status/722613351520608256/photo/1',
        5096.0, 1705.0, 12.0, 10.0, 'floof', nan, nan, nan,
        'https://pbs.twimg.com/media/Cgc9AjMVIAERdUA.jpg', 1.0,
        'Labrador_retriever', 0.5309149999999999, True,
        'golden_retriever', 0.28823000000000004, True, 'chow',
        0.044853699999999996, True]], dtype=object)

In [91]:
picture_df.shape

(2170, 28)

In [92]:
picture_df[picture_df.jpg_url.str.strip('https://') != picture_df.dog_picture_url.str.strip('http://')][['jpg_url','dog_picture_url']]

jpg_url  \
4     https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
7     https://pbs.twimg.com/media/DFyBahAVwAAhUTd.jpg   
18    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
22    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
25    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
29                                                NaN   
33                                                NaN   
34    https://pbs.twimg.com/media/DEi_N9qXYAAgEEw.jpg   
39                                                NaN   
44    https://pbs.twimg.com/media/DEF2-_hXoAAs62q.jpg   
52                                                NaN   
60    https://pbs.twimg.com/media/DDXmPrbWAAEKMvy.jpg   
61                                                NaN   
67    https://pbs.twimg.com/media/DDIKMXzW0AEibje.jpg   
68                                                NaN   
71    https://pbs.twimg.com/media/DC5YqoQW0AArOLH.jpg   
76                                                NaN   
81                                                NaN   
82    https://pbs.twimg.com/media/DCSzF3NVoAAPzT4.jpg   
86    https://pbs.twimg.com/media/DCEeLxjXsAAvNSM.jpg   
89    https://pbs.twimg.com/media/DB1m871XUAAw5vZ.jpg   
90    https://pbs.twimg.com/media/DBzhx0PWAAEhl0E.jpg   
95    https://pbs.twimg.com/media/DBkfY58XcAEdzZy.jpg   
96    https://pbs.twimg.com/media/DBg_HT9WAAEeIMM.jpg   
97                                                NaN   
100                                               NaN   
101   https://pbs.twimg.com/media/DBUxSSTXsAA-Jn1.jpg   
103   https://pbs.twimg.com/media/DBP1asiUAAEKZI5.jpg   
113   https://pbs.twimg.com/media/DAr0tDZXUAEMvdu.jpg   
116                                               NaN   
...                                               ...   
1634                                              NaN   
1635                                              NaN   
1642  https://pbs.twimg.com/media/CWMyl9EWUAAnZJ0.jpg   
1648  https://pbs.twimg.com/media/CWJQ4UmWoAIJ29t.jpg   
1650                                              NaN   
1660                                              NaN   
1667  https://pbs.twimg.com/media/CWCbd8ZWoAAtqoH.jpg   
1682  https://pbs.twimg.com/media/CV9SrABU4AQI46z.jpg   
1683  https://pbs.twimg.com/media/CV9EvZNUwAAgLCK.jpg   
1707  https://pbs.twimg.com/media/CV0_BSuWIAIvE9k.jpg   
1709  https://pbs.twimg.com/media/CV0zkzEU4AAzLc5.jpg   
1711                                              NaN   
1721                                              NaN   
1722  https://pbs.twimg.com/media/CVwx3dQXAAA0ksL.jpg   
1730                                              NaN   
1732                                              NaN   
1755  https://pbs.twimg.com/media/CVoha_IU4AAZ7vi.jpg   
1756                                              NaN   
1784  https://pbs.twimg.com/media/CVgdFjNWEAAxmbq.jpg   
1785  https://pbs.twimg.com/media/CVgbIobUYAEaeI3.jpg   
1821  https://pbs.twimg.com/media/CVRkLuJWUAAhhYp.jpg   
1847  https://pbs.twimg.com/media/CVKZsHtWwAA6gPj.jpg   
1851  https://pbs.twimg.com/media/CVJ79MzW4AEpTom.jpg   
1854                                              NaN   
1855  https://pbs.twimg.com/media/CVHRIiqWEAAj98K.jpg   
1965                                              NaN   
2005                                              NaN   
2025  https://pbs.twimg.com/media/CUdtP1xUYAIeBnE.jpg   
2028                                              NaN   
2112                                              NaN   

                                     dog_picture_url  
4     http://pbs.twimg.com/media/DF6hr6AVYAAZ8G8.jpg  
7     http://pbs.twimg.com/media/DFyBag_UQAAhhBC.jpg  
18    http://pbs.twimg.com/media/DFTH_OiUMAE-k4M.jpg  
22    http://pbs.twimg.com/media/DFDw2tsUAAEw7XW.jpg  
25    http://pbs.twimg.com/media/DE8yicKXoAAnSF8.jpg  
29                                              None  
33                                              None  
34    http://pbs.twimg.com/media/DEi_N9pXsAERvps.jpg  
39              

#### Conclusion -- there are too many different pictures used (probably because a tweet contained more than one pic), therefore will keep both ` dog_picture_url` in df_tweet and `jpg_url` in df_class, whereby  ` dog_picture_url` will refer to one of the dog's pictures, and  `jpg_url` will refer to a specific dog picture used in a classification algorithm

#### Rename columns in `class_clean` for easier comrehension

In [93]:
class_clean.columns

Index(['tweet_id', 'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2',
       'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'],
      dtype='object')

In [94]:
class_clean.sample().values

array([[750086836815486976,
        'https://pbs.twimg.com/media/Cmf5WLGWYAAcmRw.jpg', 1, 'pug',
        0.9782770000000001, True, 'teddy', 0.00313446, False,
        'Brabancon_griffon', 0.00306149, True]], dtype=object)

In [95]:
# rename jpg_url to better represent its meaning
class_clean.rename(columns={'jpg_url':'dog_picture_used_for_classification',
                            'img_num':'total_image_number',
                            'p1': 'breed_alt1',
                            'p1_conf': 'probability_alt1',
                            'p1_dog': 'classified_as_dog_alt1',
                            'p2':'breed_alt2',
                            'p2_conf':'probability_alt2',
                            'p2_dog':'classified_as_dog_alt2',
                            'p3':'breed_alt3',
                            'p3_conf':'probability_alt3',
                            'p3_dog':'classified_as_dog_alt3',}, inplace=True)

In [96]:
class_clean.columns

Index(['tweet_id', 'dog_picture_used_for_classification', 'total_image_number',
       'breed_alt1', 'probability_alt1', 'classified_as_dog_alt1',
       'breed_alt2', 'probability_alt2', 'classified_as_dog_alt2',
       'breed_alt3', 'probability_alt3', 'classified_as_dog_alt3'],
      dtype='object')

In [97]:
class_clean.breed_alt1.value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
malamute                      30
cocker_spaniel                30
French_bulldog                26
miniature_pinscher            23
Chesapeake_Bay_retriever      23
seat_belt                     22
German_shepherd               20
Staffordshire_bullterrier     20
Siberian_husky                20
Cardigan                      19
web_site                      19
beagle                        18
Eskimo_dog                    18
Shetland_sheepdog             18
Maltese_dog                   18
teddy                         18
Shih-Tzu                      17
Lakeland_terrier              17
Rottweiler                    17
kuvasz                        16
Italian_greyhound             16
          

In [98]:
class_clean[class_clean.breed_alt1 == 'web_site']

tweet_id              dog_picture_used_for_classification  \
94    667550882905632768  https://pbs.twimg.com/media/CUObvUJVEAAnYPF.jpg   
95    667550904950915073  https://pbs.twimg.com/media/CUOb_gUUkAACXdS.jpg   
130   668291999406125056  https://pbs.twimg.com/media/CUZABzGW4AE5F0k.jpg   
213   670040295598354432  https://pbs.twimg.com/media/CUx2F6lVEAAvFev.jpg   
270   670822709593571328  https://pbs.twimg.com/media/CU89schWIAIHQmA.jpg   
715   685547936038666240  https://pbs.twimg.com/media/CYOONfZW8AA7IOA.jpg   
960   705786532653883392  https://pbs.twimg.com/media/Cct1G6vVAAI9ZjF.jpg   
1210  742465774154047488  https://pbs.twimg.com/media/Ck3EribXEAAPhZn.jpg   
1317  755206590534418437  https://pbs.twimg.com/media/CnsIT0WWcAAul8V.jpg   
1443  775729183532220416  https://pbs.twimg.com/media/CsPxk85XEAAeMQj.jpg   
1447  776088319444877312  https://pbs.twimg.com/media/CsU4NKkW8AUI5eG.jpg   
1549  792913359805018113  https://pbs.twimg.com/media/CwD-eCLWIAA6v0B.jpg   
1626  804026241225523202  https://pbs.twimg.com/media/Cyh5mQTW8AQpB6K.jpg   
1696  816450570814898180  https://pbs.twimg.com/media/C1SddosXUAQcVR1.jpg   
1844  838916489579200512  https://pbs.twimg.com/media/C6RkiQZUsAAM4R4.jpg   
1847  839290600511926273  https://pbs.twimg.com/media/C6XBt9XXEAEEW9U.jpg   
1853  840696689258311684  https://pbs.twimg.com/media/C6rBLenU0AAr8MN.jpg   
1900  851464819735769094  https://pbs.twimg.com/media/C9ECujZXsAAPCSM.jpg   
2026  882045870035918850  https://pbs.twimg.com/media/DD2oCl2WAAEI_4a.jpg   

      total_image_number breed_alt1  probability_alt1  classified_as_dog_alt1  \
94                     1   web_site          0.998258                   False   
95                     1   web_site          0.999335                   False   
130                    1   web_site          0.995535                   False   
213                    1   web_site          0.901552                   False   
270                    1   web_site          0.993887                   False   
715                    1   web_site          0.923987                   False   
960                    1   web_site          0.550294                   False   
1210                   1   web_site          0.997154                   False   
1317                   1   web_site          0.906673                   False   
1443                   1   web_site          0.989407                   False   
1447                   3   web_site          0.999916                   False   
1549                   4   web_site          0.226716                   False   
1626                   1   web_site          0.492709                   False   
1696                   1   web_site          0.352857                   False   
1844                   2   web_site          0.993651                   False   
1847                   1   web_site          0.670892                   False   
1853                   1   web_site          0.841768                   False   
1900                   2   web_site          0.919649                   False   
2026                   1   web_site          0.949591                   False   

              breed_alt2  probability_alt2  classified_as_dog_alt2  \
94            dishwasher          0.000201                   False   
95                vizsla          0.000081                    True   
130                skunk          0.001363                   False   
213               borzoi          0.026660                    True   
270            Chihuahua          0.001252                    True   
715         oscilloscope          0.009712                   False   
960   Labrador_retriever          0.148496                    True   
1210          comic_book          0.000439                   False   
1317             printer          0.008600                   False   
1443  hand-held_computer          0.002139                   False   
1447                 pug          0.000077                    True   
1549          

#### Ok, not sure how good is the classification, or maybe could have chosen a different picture for classification, but will not dig much; will just lowercase all breads for clenliness

In [99]:
class_clean.breed_alt1 = class_clean.breed_alt1.str.lower()
class_clean.breed_alt2 = class_clean.breed_alt2.str.lower()
class_clean.breed_alt3 = class_clean.breed_alt3.str.lower()

### Last check and save

In [100]:
tweet_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                  timestamp  \
0 2017-08-01 16:23:56+00:00   
1 2017-08-01 00:17:27+00:00   
2 2017-07-31 00:18:03+00:00   
3 2017-07-30 15:58:51+00:00   
4 2017-07-29 16:00:24+00:00   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text      name  \
0  This is Phineas. He's a mystical boy. Only eve...   Phineas   
1  This is Tilly. She's just checking pup on you....     Tilly   
2  This is Archie. He is a rare Norwegian Pouncin...    Archie   
3  This is Darla. She commenced a snooze mid meal...     Darla   
4  This is Franklin. He would like you to stop ca...  Franklin   

                                  dog_picture_url  \
0  http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   
1  http://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2  http://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
3  http://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
4  http://pbs.twimg.com/media/DF6hr6AVYAAZ8G8.jpg   

                                       tweet_web_url  favourite_count  \
0  https://twitter.com/dog_rates/status/892420643...          37434.0   
1  https://twitter.com/dog_rates/status/892177421...          32164.0   
2  https://twitter.com/dog_rates/status/891815181...          24251.0   
3  https://twitter.com/dog_rates/status/891689557...          40731.0   
4  https://twitter.com/dog_rates/status/891327558...          38968.0   

   retweet_count  rating_numerator  rating_denominator dog_stages_alt1  \
0         8138.0              13.0                10.0             NaN   
1         6029.0              13.0                10.0             NaN   
2         3991.0              12.0                10.0             NaN   
3         8304.0              13.0                10.0             NaN   
4         8992.0              12.0                10.0             NaN   

  dog_stages_alt2 dog_stages_alt3 dog_stages_alt4  
0             NaN             NaN             NaN  
1             NaN             NaN             NaN  
2             NaN             NaN             NaN  
3             NaN             NaN             NaN  
4             NaN             NaN             NaN

In [101]:
class_clean.head()

tweet_id              dog_picture_used_for_classification  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   total_image_number              breed_alt1  probability_alt1  \
0                   1  welsh_springer_spaniel          0.465074   
1                   1                 redbone          0.506826   
2                   1         german_shepherd          0.596461   
3                   1     rhodesian_ridgeback          0.408143   
4                   1      miniature_pinscher          0.560311   

   classified_as_dog_alt1          breed_alt2  probability_alt2  \
0                    True              collie          0.156665   
1                    True  miniature_pinscher          0.074192   
2                    True            malinois          0.138584   
3                    True             redbone          0.360687   
4                    True          rottweiler          0.243682   

   classified_as_dog_alt2           breed_alt3  probability_alt3  \
0                    True    shetland_sheepdog          0.061428   
1                    True  rhodesian_ridgeback          0.072010   
2                    True           bloodhound          0.116197   
3                    True   miniature_pinscher          0.222752   
4                    True             doberman          0.154629   

   classified_as_dog_alt3  
0                    True  
1                    True  
2                    True  
3                    True  
4                    True

In [102]:
#write to file

tweet_clean.to_csv('twitter_archive_master.csv',index=False)
class_clean.to_csv('twitter_archive_master_class.csv',index=False)